### DISCUSSION TITLE
Workflow: difference between the branches

### DISCUSSION THREAD

**Discussion by MrGitGo** [community/community#21811] [New to GitHub]
<p>So I am currently integrating a workflow for our new git repository. I am a little bit confused what is done in the different branches (master,release,develop, feature):</p>
<p>Master: The master ist the main branch where you have you publish your end realses </p>
<p>Release: I guess you do just bugfixes here</p>
<p>develop and feature: Here I am not sure where how to work. You have the feature branch where you implement features and the develop also for bugfixes?</p>
<p>I am referring to this picture:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/1X/1b76f5ac8feafb5ed841a1c90c13b6e0d7115a70.png" data-download-href="/uploads/short-url/3UXLdCoGZmNwEN09jfAxLJzYxJm.png?dl=1" title="git-model@2x.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/33628476/181066985-0df26fd4-8c40-453b-b3be-538ba4e57262.png" alt="git-model@2x.png" data-base62-sha1="3UXLdCoGZmNwEN09jfAxLJzYxJm" width="377" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/1X/1b76f5ac8feafb5ed841a1c90c13b6e0d7115a70_2_377x500.png, https://aws1.discourse-cdn.com/github/optimized/1X/1b76f5ac8feafb5ed841a1c90c13b6e0d7115a70_2_565x750.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/1X/1b76f5ac8feafb5ed841a1c90c13b6e0d7115a70_2_754x1000.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/1X/1b76f5ac8feafb5ed841a1c90c13b6e0d7115a70_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">git-model@2x.png</span><span class="informations">1150×1524 125 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by DannyBen** [comment, marked as answer]
<p>I believe this post is more suitable in the “Using Git and GitHub” topic, but nonetheless - here is my advice:</p>
<p>If this flow confuses you, don’t use it. </p>
<p>I am working with git and github for many years on many projects, and this chart confuses me as well.</p>
<p>Here is what I do both for my personal projects and for my professional projects (work):</p>
<ul>
<li>
<p>My master branch always includes the most up to date code, after it was tested and in a state that is ready to be deployed to production. Master is sacred in my workflow, it should never receive new buggy code or code that failed tests.</p>
</li>
<li>
<p>Whenever I develop a new feature or a bugfix, I do it in a new branch. You can call these new branches however you wish. You can either choose a zero convention policy (so you just name them my-new-feature or whatever), or you can decide on some conventions (for example, all feature branches start with “feature/” and all bug fixes start with “fix/”. This sort of convention depends on the organization, how many people are working on this repo, and personal taste.</p>
</li>
<li>
<p>Features and bugfixes are merged back to the master branch only after they passed the acceptable tests in your organization - usually this means automated tests and sometimes, code review by someone other than the coder.</p>
</li>
<li>
<p>Once feature / bug / other branches were merged to master, they are deleted.</p>
</li>
<li>
<p>You should aim to have these branches have a short “shelf life” - work on small units of code, merge them to master when they are ready. In other words - don’t let your branches list get too long. Of course, in larger organizations, this is not always possible.</p>
</li>
<li>
<p>In some extreme cases, where I need a “long lived branch” - for example, when I am working on a long upgrade or something - then I am creating this alternative branch, creating a “Draft Pull Request” for it, and keeping it always up to date with any changes done in the master branch. </p>
</li>
<li>
<p>All merges to master are done through a Pull Reuqest on GitHub. Even if it is a piece of code that I wrote on my own repo. This way you have an organized place to review the code, and describe/discuss the feature in length if needed. And it serves as a nice “log” for everything you have done.</p>
</li>
</ul>
<p>Hope this helps.</p>
